File: leetcode/42. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left_max = 0;
        int right_max = 0;
        int i = 0;
        int j = height.size() - 1;
        int res = 0;
        while(i < j){
            while(i < j && height[i] < height[j]){
                left_max = max(left_max, height[i]);
                res += min(left_max, height[j]) - height[i];
                ++i;
            }
            
            while(i < j && height[i] >= height[j]){
                right_max = max(right_max, height[j]);
                res += min(right_max, height[i]) - height[j];
                --j;
            }
        }
        return res;
    }
};

File: leetcode/42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        i = 0
        j = len(height) - 1
        
        res = 0
        left_max = 0
        right_max = 0
        
        while i < j:
            left_max = max(left_max, height[i])
            right_max = max(right_max, height[j])
            
            while i < j and height[i] < height[j]:
                i += 1
                left_max = max(left_max, height[i])
                res += left_max - height[i]
            
            while i < j and height[i] >= height[j]:
                j -= 1
                right_max = max(right_max, height[j])
                res += right_max - height[j]
        
        return res



class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        res = 0
        for i in range(len(height)):
            h = height[i]
            while len(stack) != 0 and h >= height[stack[-1]]:
                j = stack.pop()
                if len(stack) > 0:
                    res += (min(height[stack[-1]], h) - height[j]) * (i-stack[-1]-1)
            stack.append(i)
        
        return res
